File: src/dotnet/Harness.Engine/HarnessState.cs
namespace Harness.Engine;

/// <summary>
/// State persisted across invocations: step counter + accumulated domain data.
/// Top-level type (not nested) so it's servable by System.Text.Json's source generator, a
/// Native AOT requirement.
/// </summary>
public record HarnessState(int Step, Dictionary<string, string> Data)
{
    // Run's accumulated cost, input for the cost ceiling (see TaskRegistry). init
    // properties (non-positional) so they don't break the existing `new HarnessState(0, new())`
    // calls; kept out of Data so it doesn't pollute the evaluation's completeness check.

    /// <summary>Instruction chars emitted so far — the cost proxy (sum of <c>InstructionChars</c>).</summary>
    public int CostChars { get; init; }

    /// <summary>
    /// Driver context (e.g. <c>{"driver":"claude code"}</c>) captured in the <c>start</c>
    /// envelope — survives across invocations so <see cref="PromptFormatter"/> can
    /// reinject it into every output without each task passing it along manually.
    /// </summary>
    public Dictionary<string, string>? Context { get; init; }
}

File: src/go/harnessengine/harness_state.go
package harnessengine

// HarnessState is the state persisted between invocations: step counter plus accumulated
// domain data.
type HarnessState struct {
	Step int               `json:"step"`
	Data map[string]string `json:"data"`

	// CostChars is the run's accumulated cost, input to the cost ceiling (see TaskRegistry).
	CostChars int `json:"costChars"`

	// Context is the driver context (e.g. {"driver":"claude code"}) captured in the `start`
	// envelope — survives across invocations so PromptFormatter can reinject it into every
	// output without each task having to pass it along.
	Context map[string]string `json:"context,omitempty"`

	// TerminalReason latches a hard stop across process boundaries.
	TerminalReason string `json:"terminalReason,omitempty"`
}

// NewHarnessState builds a state with the given step and data (data defaults to an empty,
// non-nil map).
func NewHarnessState(step int, data map[string]string) HarnessState {
	if data == nil {
		data = map[string]string{}
	}
	return HarnessState{Step: step, Data: data}
}
